client/lib/upgrades/actions/cart.js

```diff
@@ -138,6 +146,76 @@ export function removeCoupon() {
 	} );
 }
 
+export function saveCouponQueryArgument() {
+	// read coupon query argument, return early if there is none
+	const parsedUrl = urlParseAmpCompatible( location.href );
+	const couponCode = parsedUrl.query.coupon;
+	if ( ! couponCode ) {
+		return;
+	}
+
+	// read coupon list from localStorage, create new if it's not there yet, refresh existing
+	const couponsJson = localStorage.getItem( MARKETING_COUPONS_KEY );
+	const coupons = JSON.parse( couponsJson ) || {};
+	const ONE_WEEK_MILLISECONDS = 7 * 24 * 60 * 60 * 1000;
+	const now = Date.now();
+	debug( 'Found coupons in localStorage: ', coupons );
+
+	coupons[ couponCode ] = now;
+
+	// delete coupons if they're older than a week
+	Object.keys( coupons ).forEach( key => {
+		if ( now > coupons[ key ] + ONE_WEEK_MILLISECONDS ) {
+			delete coupons[ key ];
+		}
+	} );
+
+	// write remembered coupons back to localStorage
+	debug( 'Storing coupons in localStorage: ', coupons );
+	localStorage.setItem( MARKETING_COUPONS_KEY, JSON.stringify( coupons ) );
+}
+
+export function getRememberedCoupon() {
+	// read coupon list from localStorage, return early if it's not there
+	const couponsJson = localStorage.getItem( MARKETING_COUPONS_KEY );
+	const coupons = JSON.parse( couponsJson );
+	if ( ! coupons ) {
+		debug( 'No coupons found in localStorage: ', coupons );
+		return null;
+	}
+
+	const COUPON_CODE_WHITELIST = [ 'LEIFTEST', 'PATREON' ];
```

Comment:
Was it intentional to leave 'LEIFTEST' here in the whitelist array?

Reply:
Whoops! No. Removed.